api/src/main/java/com/github/retrooper/packetevents/protocol/world/chunk/palette/ListPalette.java

```
@@ -63,6 +64,12 @@ public ListPalette(int bitsPerEntry, PacketWrapper<?> wrapper) {
         this.nextId = paletteLength;
     }
 
+    public ListPalette(int bitsPerEntry, int[] data) {
+        this.bits = bitsPerEntry;
+        this.data = data;
```

please verify the length of the data matches the bits-per-entry parameter

---

I'm pretty sure the data array length has to exactly match   the bits, though the array doesn't need to be fully filled

---

I would do a `Arrays.copyOf(data, expectedSize)` if the data array length is lower than the bits, does it make sense or I should throw an exception?

---

Copying the array is fine :+1:
But if you do that, please always copy/clone it, so the constructor behaves consistently

---

It's the behavior now.